feat(access): Implement API for changing team roles

RyanSkonnord opened the pull request:
Broken out from https://github.com/getsentry/sentry/pull/32498. Follows https://github.com/getsentry/sentry/pull/33387.

    Add a PUT method to OrganizationMemberTeamDetailsEndpoint for assigning
    team-based roles, applying permission checks analogously to assignment
    of organization-based roles. Add a corresponding GET method, currently
    unused, but as a RESTful best practice.

    Modify organization-based role assignment so that, if a member is given
    an organization-level promotion with a new mimumum team role that
    supersedes their assigned team role, the assigned team role is nulled
    out so that it doesn't persist invisibly.

RyanSkonnord commented on src/sentry/api/endpoints/organization_member_team_details.py:
```diff
@@ -103,6 +123,23 @@ def _create_access_request(
+        if not OrganizationMemberTeam.objects.filter(team=team, organizationmember=member).exists():
+            raise ResourceDoesNotExist
+
+        return Response(serialize(team, request.user, TeamWithProjectsSerializer()), status=200)
```
Following up on https://github.com/getsentry/sentry/pull/32498#discussion_r833342190 from @markstory:

> Do we need the `projects` in this response? I thought we were trying to move away from automatically including 1:N relations.

In practical terms, no. I believe only the Team Members view uses this endpoint, and when I replace all instances of `TeamWithProjectsSerializer` with `TeamSerializer`, the code that calls the existing POST and DELETE methods continues to work fine. The endpoint is undocumented, so I think there's no abstract harm in changing it.

However, I could use some advice on front-end best practices before I make the change. The [code in `teamMembers.tsx`](https://github.com/getsentry/sentry/blob/bb42837000875ecebc4276a0d6b4a68b30375eb4/static/app/views/settings/organizationTeams/teamMembers.tsx#L49) that unpacks the response does so with the `Member` type, which [has `projects` as an attribute](https://github.com/getsentry/sentry/blob/496e21347d491f2d06bc0b6eac5a08c9b2ec00a5/static/app/types/organization.tsx#L102).

If we remove projects from the response, should we also change the front-end type? Is it worth it?

nhsiehgit replied:
without understanding fully what this serializer does, 

I'd assume yes - you would probably want to remove projects from the types file.
TS would shout at you and help point out if anythings depending on that type being available (and likely suggest you to clean up some FE flows if there are)

If there are dependencies, probably good to remove the FE deps first and ensure nothings expecting `projects`, then remove it from the api response 🤷

markstory replied:
Agree with Nathan that removing the front-end attribute from the types first will help identify any UI flows that need to change, ship those changes and then deploy the endpoint changes.

RyanSkonnord replied:
👍 Will try removing projects from the type.

I also want to see if we can simplify the whole thing by serializing the `OrganizationMemberTeam` rather than the entire `Team`. I believe `teamMembers.tsx` is the only thing consuming the POST response and I don't think it needs all of it.

RyanSkonnord replied:
It dawned on me that this endpoint used to be only a POST method and was reflecting the state of the parent object (the team). Since GET and PUT are brand new, I was making a mistake by trying to have their response contract match POST. (Further complicating things was the DELETE body, which is just unnecessary.)

As of 11bc4a1fd74c0d7b5fd8b93b2e436ce645035bdb, GET and PUT use the newly introduced `OrganizationMemberTeamDetailsSerializer`. It's very minimal and we can always add more.

It might still be worth taking the projects out of the POST response, but that's a preexisting issue that we can handle separately.

nhsiehgit reviewed (approved):
👍

RyanSkonnord committed:
feat(access): Implement API for changing team roles

Add a PUT method to OrganizationMemberTeamDetailsEndpoint for assigning
team-based roles, applying permission checks analogously to assignment
of organization-based roles. Add a corresponding GET method, currently
unused, but as a RESTful best practice.

Modify organization-based role assignment so that, if a member is given
an organization-level promotion with a new mimumum team role that
supersedes their assigned team role, the assigned team role is nulled
out so that it doesn't persist invisibly.

nhsiehgit reviewed (approved):
LGTM

RyanSkonnord committed:
Optimize GET query; add unit test coverage

RyanSkonnord committed:
Revert unneeded changes from squash merge

RyanSkonnord committed:
Optimize membership query in TeamSerializer

Satisfies OrganizationDetailsTest.test_with_projects without increasing
number of database interactions.

RyanSkonnord committed:
Merge branch 'master' into team-role-api

markstory reviewed (approved):
Looks good to me.

RyanSkonnord committed:
Oops, needed that after all

This reverts commit 39e205ed5a45011d444c911bbc4ba79f9d1f2e01.

RyanSkonnord committed:
Rename TeamSerializer.role field to teamRole

RyanSkonnord committed:
Add assertions for teamRole

RyanSkonnord committed:
Rename role to teamRole on OrganizationMemberTeamDetailsEndpoint

RyanSkonnord committed:
Fix TeamSerializerResponse type

RyanSkonnord committed:
Merge branch 'master' into team-role-api
